File: orchestration/src/tasks/helpers/motion.py
from __future__ import annotations

from pathlib import Path
from typing import Any

from src.configs.run_config import RunConfig
from src.tasks.helpers.artifacts import file_sha256, write_text
from src.tasks.helpers.rosbag_profiles import profile_topics
from src.tasks.helpers.slam_hover import (
    build_p6_doctor_summary,
)

P7_ROSBAG_CONTAINER = "navlab-p7-rosbag"

def build_p7_doctor_summary(
    config: RunConfig,
    *,
    runtime_config: Path,
    include_dependencies: bool = True,
) -> dict[str, Any]:
    p7 = config.orchestration.motion_gate
    p6_runtime = config.artifact_dir / "p7_doctor_p6_slam_hover_runtime.toml"
    p6_doctor = (
        build_p6_doctor_summary(config, runtime_config=p6_runtime)
        if include_dependencies
        else {"ok": True, "blockers": [], "skipped": "acceptance already launched P6 prerequisites"}
    )
    profile_path = Path(p7.rosbag_profile)
    required, optional, topics = profile_topics(profile_path)
    blockers = [str(item) for item in p6_doctor.get("blockers", [])]
    if not profile_path.is_file() or not topics:
        blockers.append("P7 rosbag profile is missing or empty")
    if p7.uses_gazebo_truth_as_input:
        blockers.append("P7 must not use Gazebo truth as a control/planning/SLAM/ExternalNav input")
    if p7.slam_odom_topic != config.orchestration.slam_backend.slam_odom_topic:
        blockers.append("P7 SLAM odom topic must match P3 canonical SLAM odom topic")
    if p7.slam_odom_topic == p7.truth_diagnostic_topic:
        blockers.append("P7 SLAM odom topic must not be the Gazebo truth diagnostic topic")
    if p7.cmd_vel_topic != config.orchestration.fcu_controller.cmd_vel_topic:
        blockers.append("P7 cmd_vel topic must match the P4 FCU controller output topic")
    return {
        "ok": not blockers,
        "blocked": bool(blockers),
        "blockers": blockers,
        "p7_motion_gate_doctor": {
            "runtime_config": str(runtime_config),
            "runtime_config_sha256": file_sha256(runtime_config) if runtime_config.is_file() else "",
            "dependency_checks_included": include_dependencies,
            "slam_odom_topic": p7.slam_odom_topic,
            "external_nav_status_topic": p7.external_nav_status_topic,
            "motion_status_topic": p7.motion_status_topic,
            "uses_gazebo_truth_as_input": p7.uses_gazebo_truth_as_input,
            "hover_claim": p7.hover_claim,
            "motion_claim": p7.motion_claim,
            "exploration_claim": p7.exploration_claim,
            "thresholds": {
                "motion_distance_m": p7.motion_distance_m,
                "motion_speed_mps": p7.motion_speed_mps,
                "yaw_scan_rad": p7.yaw_scan_rad,
                "max_stop_drift_m": p7.max_stop_drift_m,
                "min_clearance_m": p7.min_clearance_m,
            },
            "rosbag_profile": {
                "profile": str(profile_path),
                "required_topics": required,
                "optional_topics": optional,
            },
        },
        "p6_slam_hover_doctor": p6_doctor,
    }


def _write_foxglove_notes(config: RunConfig) -> None:
    p7 = config.orchestration.motion_gate
    write_text(
        config.artifact_dir / "foxglove_notes.md",
        "\n".join(
            [
                "# NavLab P7 motion gate replay notes",
                "",
                "P7 validates forward/back/yaw/stop motion after real SLAM hover. It is not an exploration gate.",
                "",
                "- Fixed frame: `map`.",
                f"- Motion status: `{p7.motion_status_topic}`.",
                f"- Setpoint output: `{p7.setpoint_output_topic}`.",
                f"- SLAM odom: `{p7.slam_odom_topic}`.",
                f"- FCU pose/twist: `{p7.fcu_pose_topic}`, `{p7.fcu_twist_topic}`.",
                f"- Diagnostic truth only: `{p7.truth_diagnostic_topic}`.",
                "- Do not use Gazebo truth as a SLAM, ExternalNav, planning, or control input.",
            ]
        )
        + "\n",
    )

File: orchestration/sim/internal/tasks/helpers/motion.go
package helpers

import (
	"crypto/sha256"
	"encoding/hex"
	"os"
)

const MotionRosbagContainer = "navlab-motion-rosbag"

type MotionGateSpec struct {
	RuntimeConfigPath      string
	RosbagProfile          string
	SlamOdomTopic          string
	CanonicalSlamOdomTopic string
	ExternalNavStatusTopic string
	MotionStatusTopic      string
	CmdVelTopic            string
	CanonicalCmdVelTopic   string
	SetpointOutputTopic    string
	FCUPoseTopic           string
	FCUTwistTopic          string
	TruthDiagnosticTopic   string
	UsesGazeboTruthAsInput bool
	HoverClaim             string
	MotionClaim            string
	ExplorationClaim       string
	MotionDistanceM        float64
	MotionSpeedMPS         float64
	YawScanRad             float64
	MaxStopDriftM          float64
	MinClearanceM          float64
}

type DependencyDoctor struct {
	OK       bool     `json:"ok"`
	Blockers []string `json:"blockers"`
	Skipped  string   `json:"skipped,omitempty"`
}

type MotionDoctorSummary struct {
	OK                    bool             `json:"ok"`
	Blocked               bool             `json:"blocked"`
	Blockers              []string         `json:"blockers"`
	MotionGateDoctor      MotionGateDoctor `json:"motion_gate_doctor"`
	HoverDependencyDoctor DependencyDoctor `json:"hover_dependency_doctor"`
}

type MotionGateDoctor struct {
	RuntimeConfig            string              `json:"runtime_config"`
	RuntimeConfigSHA256      string              `json:"runtime_config_sha256"`
	DependencyChecksIncluded bool                `json:"dependency_checks_included"`
	SlamOdomTopic            string              `json:"slam_odom_topic"`
	ExternalNavStatusTopic   string              `json:"external_nav_status_topic"`
	MotionStatusTopic        string              `json:"motion_status_topic"`
	UsesGazeboTruthAsInput   bool                `json:"uses_gazebo_truth_as_input"`
	HoverClaim               string              `json:"hover_claim"`
	MotionClaim              string              `json:"motion_claim"`
	ExplorationClaim         string              `json:"exploration_claim"`
	Thresholds               MotionThresholds    `json:"thresholds"`
	RosbagProfile            MotionRosbagProfile `json:"rosbag_profile"`
}

type MotionThresholds struct {
	MotionDistanceM float64 `json:"motion_distance_m"`
	MotionSpeedMPS  float64 `json:"motion_speed_mps"`
	YawScanRad      float64 `json:"yaw_scan_rad"`
	MaxStopDriftM   float64 `json:"max_stop_drift_m"`
	MinClearanceM   float64 `json:"min_clearance_m"`
}

type MotionRosbagProfile struct {
	Profile        string   `json:"profile"`
	RequiredTopics []string `json:"required_topics"`
	OptionalTopics []string `json:"optional_topics"`
}

func DefaultMotionGateSpec() MotionGateSpec {
	return MotionGateSpec{
		RuntimeConfigPath:      "motion_gate_runtime.toml",
		RosbagProfile:          "",
		SlamOdomTopic:          "/slam/odom",
		CanonicalSlamOdomTopic: "/slam/odom",
		ExternalNavStatusTopic: "/external_nav/status",
		MotionStatusTopic:      "/navlab/motion/status",
		CmdVelTopic:            "/ap/cmd_vel",
		CanonicalCmdVelTopic:   "/ap/cmd_vel",
		SetpointOutputTopic:    "/navlab/setpoint/output",
		FCUPoseTopic:           "/ap/pose/filtered",
		FCUTwistTopic:          "/ap/twist/filtered",
		TruthDiagnosticTopic:   "/navlab/truth/odom",
		UsesGazeboTruthAsInput: false,
		HoverClaim:             "evaluated",
		MotionClaim:            "evaluated",
		ExplorationClaim:       "not_evaluated",
		MotionDistanceM:        0.5,
		MotionSpeedMPS:         0.1,
		YawScanRad:             0.6,
		MaxStopDriftM:          0.2,
		MinClearanceM:          0.35,
	}
}

func BuildMotionDoctorSummary(spec MotionGateSpec, dependency DependencyDoctor, includeDependencies bool) MotionDoctorSummary {
	if spec.RuntimeConfigPath == "" {
		spec = DefaultMotionGateSpec()
	}
	if !includeDependencies {
		dependency = DependencyDoctor{
			OK:       true,
			Blockers: []string{},
			Skipped:  "acceptance already launched hover prerequisites",
		}
	}
	blockers := append([]string{}, dependency.Blockers...)
	topics := Topics{}
	if spec.RosbagProfile == "" {
		blockers = append(blockers, "motion rosbag profile is missing or empty")
	} else if loaded, err := ProfileTopics(spec.RosbagProfile); err != nil || len(loaded.All) == 0 {
		blockers = append(blockers, "motion rosbag profile is missing or empty")
	} else {
		topics = loaded
	}
	if spec.UsesGazeboTruthAsInput {
		blockers = append(blockers, "motion gate must not use Gazebo truth as a control/planning/SLAM/ExternalNav input")
	}
	if spec.SlamOdomTopic != spec.CanonicalSlamOdomTopic {
		blockers = append(blockers, "motion gate SLAM odom topic must match canonical SLAM odom topic")
	}
	if spec.SlamOdomTopic == spec.TruthDiagnosticTopic {
		blockers = append(blockers, "motion gate SLAM odom topic must not be the Gazebo truth diagnostic topic")
	}
	if spec.CmdVelTopic != spec.CanonicalCmdVelTopic {
		blockers = append(blockers, "motion gate cmd_vel topic must match the FCU controller output topic")
	}
	blockers = uniqueSorted(blockers)
	return MotionDoctorSummary{
		OK:       len(blockers) == 0,
		Blocked:  len(blockers) > 0,
		Blockers: blockers,
		MotionGateDoctor: MotionGateDoctor{
			RuntimeConfig:            spec.RuntimeConfigPath,
			RuntimeConfigSHA256:      fileHashIfExists(spec.RuntimeConfigPath),
			DependencyChecksIncluded: includeDependencies,
			SlamOdomTopic:            spec.SlamOdomTopic,
			ExternalNavStatusTopic:   spec.ExternalNavStatusTopic,
			MotionStatusTopic:        spec.MotionStatusTopic,
			UsesGazeboTruthAsInput:   spec.UsesGazeboTruthAsInput,
			HoverClaim:               spec.HoverClaim,
			MotionClaim:              spec.MotionClaim,
			ExplorationClaim:         spec.ExplorationClaim,
			Thresholds: MotionThresholds{
				MotionDistanceM: spec.MotionDistanceM,
				MotionSpeedMPS:  spec.MotionSpeedMPS,
				YawScanRad:      spec.YawScanRad,
				MaxStopDriftM:   spec.MaxStopDriftM,
				MinClearanceM:   spec.MinClearanceM,
			},
			RosbagProfile: MotionRosbagProfile{
				Profile:        spec.RosbagProfile,
				RequiredTopics: append([]string{}, topics.Required...),
				OptionalTopics: append([]string{}, topics.Optional...),
			},
		},
		HoverDependencyDoctor: dependency,
	}
}

func MotionFoxgloveNotes(spec MotionGateSpec) string {
	return "# NavLab motion gate replay notes\n\n" +
		"The motion gate validates forward/back/yaw/stop motion after SLAM hover. It is not an exploration gate.\n\n" +
		"- Fixed frame: `map`.\n" +
		"- Motion status: `" + spec.MotionStatusTopic + "`.\n" +
		"- Setpoint output: `" + spec.SetpointOutputTopic + "`.\n" +
		"- SLAM odom: `" + spec.SlamOdomTopic + "`.\n" +
		"- FCU pose/twist: `" + spec.FCUPoseTopic + "`, `" + spec.FCUTwistTopic + "`.\n" +
		"- Diagnostic truth only: `" + spec.TruthDiagnosticTopic + "`.\n" +
		"- Do not use Gazebo truth as a SLAM, ExternalNav, planning, or control input.\n"
}

func fileHashIfExists(path string) string {
	if path == "" {
		return ""
	}
	data, err := os.ReadFile(path)
	if err != nil {
		return ""
	}
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}
